### DISCUSSION TITLE
Sometimes I'd like that puppet does something if a certain file is on the system.

### DISCUSSION THREAD

**Discussion by pier4r** [puppetlabs/community#10] [Q&A]
Countless times I wanted to know if a file was there, as a switch for certain resources.

```
### file exists with external facts
$file_to_check_fact_path = '/etc/facter/facts.d/file_exists_itself.sh'
$file_to_check = $file_to_check_fact_path #check itself

file { $file_to_check_fact_path:
	ensure  => 'file',
	mode    => 'a+x',
	content =>"#!/bin/bash
# cat test.sh
files_to_check_arr=( '${file_to_check_fact_path}' )

for file_to_check in \${files_to_check_arr[@]}; do
test -f \"\${file_to_check}\" && {
	echo \"\${file_to_check}=present\"
} || {
	echo \"\${file_to_check}=absent\"
}
done
	",
}

notify { 'test_external_fact': message => $facts[$file_to_check_fact_path] }
```

This may be suboptimal, but at least works. Is there anything better?

**Comment by bitvijays** [comment]
@pier4r Hope you are doing well. Maybe have a look at [Exec OnlyIf](https://puppet.com/docs/puppet/7.5/types/exec.html#exec-attribute-onlyif ) and [Exec Unless](https://puppet.com/docs/puppet/7.5/types/exec.html#exec-attribute-unless)

**Reply by pier4r** [reply]
Hi! I know those. Anyway with an exec I cannot say "deploy those resources (say a firewall rule) only if this file exists", or at least I cannot see how I can build it.

For this a fact in this case helps as I can then treat it as a variable.

**Reply by swells-ac** [reply]
I am not sure if you are just talking about random files.  You can setup your hosts to have files that contain a role or something else.  Than map that by using custom facts and hierdata.  So say have a file that has "testing" in it.  Then you can map that back to have separate environment for it.

**Comment by XMol** [comment]
Hi @pier4r,

I think it is safe to say that facts are your best option in this case. Especially when your Puppet client isn't also a master and compiles the catalog for itself, the master cannot know whether files exist on the client without facts.

That is a direct consequence from the [catalog compilation process](https://puppet.com/docs/puppet/6/subsystem_catalog_compilation.html#subsystem_catalog_compilation). Before the master compiles a catalog, it gathers all external data. The sole option to react dynamically during catalog compilation are (Puppet) functions, which are run on the Puppet master. In theory, you could implement a function that contacts your client node to check whether a specific file exists. That sounds like a bad idea to me because of all the requirements attached to that. With enough determination however...

Ciao,  
Xavier.

**Comment by binford2k** [comment]
I have a module that lets you do partial catalog enforcement based on conditions you set, `binford2k-assert`. That said, _in general_ you don't want your configuration to be indeterminate. You want to specify all the conditions you care about. In other words, rather than checking to see if a file exists and then reacting appropriately, your code should specify whether the file should exist in the first place.

**Reply by pier4r** [reply]
This is true, in an ideal world we should determine things beforehand.

Anyway in less ideal world (that is most likely what happens in many cases), things are a bit more complicated.
Example: puppet apply based execution that requires modules to be installed (locally). Some part of the code requires functions in the yet-to-be-installed modules. It is not really possible (for my knowledge) to put a "care about this piece of code only when the module is there" in puppet without a facts that enables the piece of code only if the module is installed (i.e: a file exists).

If there more cleaner way, I am all ears!

Last but not least: in general I think that while one should know what is best, it should be possible to encourage also "what is possible with the tool X" type of discussions.

**Reply by binford2k** [reply]
> It is not really possible (for my knowledge) to put a "care about this piece of code only when the module is there" in puppet

There is though. `r10k puppetfile install && puppet apply manifests/site.pp`

Instead of making your code reactive to the various states that your codebase can be in, just ensure codebase consistency before enforcing it.

**Reply by pier4r** [reply]
"Instead of making your code reactive to the various states that your codebase can be in, just ensure codebase consistency before enforcing it."

Yes I agree, if one has the knowledge how to piece things together.

Interesting, I didn't think to deploy r10k together with puppet apply. I was thinking puppet should do everything on its own (otherwise I should wrap things with script, more or less always)
